packages/gatsby-graphql-source-toolkit/README.md

```diff
@@ -0,0 +1,1057 @@
+- [Tools Reference](#tools-reference)
+  - [Configuration Tools](#configuration-tools)
+    - [createDefaultQueryExecutor](#createdefaultqueryexecutor)
+    - [loadSchema](#loadschema)
+    - [buildNodeDefinitions](#buildnodedefinitions)
+  - [Query compilation tools](#query-compilation-tools)
+    - [generateDefaultFragments](#generatedefaultfragments)
+    - [readOrGenerateDefaultFragments](#readorgeneratedefaultfragments)
+    - [compileNodeQueries](#compilenodequeries)
+  - [Schema customization tools](#schema-customization-tools)
+    - [createSchemaCustomization](#createschemacustomization)
+  - [Source nodes tools](#source-nodes-tools)
+    - [sourceAllNodes](#sourceallnodes)
+    - [sourceNodeChanges](#sourcenodechanges)
+
+## Why not `gatsby-source-graphql`?
+
+Historically Gatsby suggested [`gatsby-source-graphql`][2] plugin to consume data from remote GraphQL APIs.
+
+This plugin is easy to use, but it has a significant problem: it doesn't adhere to the original
+Gatsby architecture (doesn't [source nodes][3]), which makes data caching impossible.
+As a result, it doesn't scale well, and can't work with Gatsby Preview or Incremental Builds by design
+([more technical details][4]).
+
+Also, with `gatsby-source-graphql` you can't leverage the power of Gatsby transformer plugins like `gatsby-transformer-remark`
+or `gatsby-transformer-sharp` (and it's hard to use with `gatsby-image` as a consequence).
+
+This new toolkit should solve all those issues and implement correct node sourcing for Gatsby.
+
+## Features
+
+- Efficient concurrent data fetching
+- Automatic data pagination
+- Cache data between runs (supports sourcing delta changes)
+- Customize what is sourced
+- Schema customization out of the box (no performance penalty of type inference)
+- Designed to support [Gatsby Preview][5] and [Incremental Builds][6]
+
+## How it works
+
+Let's imagine we have a GraphQL API located at `https://www.example.com/graphql`.
+This example API has the following straightforward schema:
+
+```graphql
+type Post {
+  id: ID!
+  description(truncateAt: Int): String
+  author: Author
+}
+
+type Author {
+  id: ID!
+  name: String
+  allPosts: [Post]
+}
+
+type Query {
+  author(id: ID!): Author
+  authors(limit: Int = 10, offset: Int = 0): [Author]
+  post(id: ID!): Post
+  posts(limit: Int = 10, offset: Int = 0): [Post]
+}
+```
+
+How do we source data from this GraphQL API using the toolkit?
+Let's look at the full example and then walk through it step-by-step.
+
+```js
+async function createSourcingConfig(gatsbyApi) {
+  // Step1. Setup remote schema:
+  const execute = createDefaultQueryExecutor(`https://www.example.com/graphql`)
+  const schema = await loadSchema(execute)
+
+  // Step2. Configure Gatsby node types
+  const gatsbyNodeTypes = [
+    {
+      remoteTypeName: `Post`,
+      remoteIdFields: [`__typename`, `id`],
+      queries: `query LIST_POSTS { posts(limit: $limit, offset: $offset) }`,
```

Comment:
Actually I was moving back and forth with it :) And here are the main reasons why a single document:

1. People may have multiple `LIST` queries for the same node type (maybe, with different filters)
2. There may be custom queries for special cases (e.g. Tyler uses separate queries for Previews in WP plugin).

So we can end up with many documents for a single node type. Having a single document allows us to inject fragments to all queries (even custom) and also have a unified way to run those (by just providing `operationName`).

Convention for operation names is also useful for runtime debugging and error reporting: you can simply output operation name with a list of variables.

It can also help to batch only given "kind" of queries. For example, I've found batching `nodeQuery` to make the most sense (batching `listQuery` makes much less sense). I guess we could differentiate queries by some other method but naming conventions seem to be the easiest way.

Also, IDEs and GraphQL editors support documents with multiple operations just fine and graphql-js designed in a way that supports them well too.

And there is an easy way to convert single document to a list of documents via [`separateOperations` utility](https://github.com/graphql/graphql-js/blob/master/src/utilities/separateOperations.js). It's a bit more complex to do the other way around.